### PR TITLE
bpo-34022: Stop forcing of hash-based invalidation with SOURCE_DATE_EPOCH

### PR DESCRIPTION
**Pull request by elprans**
Unconditional forcing of ``CHECKED_HASH`` invalidation was introduced in
3.7.0 in [bpo-29708](https://www.bugs.python.org/issue29708).  The change is bad, as it unconditionally overrides
*invalidation_mode*, even if it was passed as an explicit argument to
``py_compile.compile()`` or ``compileall``.  An environment variable
should *never* override an explicit argument to a library function.
That change leads to multiple test failures if the ``SOURCE_DATE_EPOCH``
environment variable is set.

This changes ``py_compile.compile()`` to only look at
``SOURCE_DATE_EPOCH`` if no explicit *invalidation_mode* was specified.
I also made various relevant tests run with explicit control over the
value of ``SOURCE_DATE_EPOCH``.

While looking at this, I noticed that ``zipimport`` does not work
with hash-based .pycs _at all_, though I left the fixes for
subsequent commits.

https://bugs.python.org/issue34022

### CODE REVIEW

File: Doc/library/py_compile.rst
Comment (merwok) [inline Doc/library/py_compile.rst]: its

File: Doc/library/compileall.rst
Comment (vstinner) [inline Doc/library/compileall.rst:97]: "is set": I suggest "is set to a non-empty string". See other existing examples:
https://docs.python.org/dev/using/cmdline.html#envvar-PYTHONOPTIMIZE

Maybe add a reference to the PEP 552 somewhere (you can use syntax :pep:`552`).

By the way, should we mention SOURCE_DATE_EPOCH somewhere in https://docs.python.org/dev/using/cmdline.html ? The variable is not specific to Python, but it changes its behavior.

File: Doc/library/py_compile.rst
Comment (vstinner) [inline Doc/library/py_compile.rst:60]: Maybe add a reference to the PEP 552 somewhere (you can use syntax :pep:`552`).

File: Lib/test/test_py_compile.py
Comment (vstinner) [inline Lib/test/test_py_compile.py:19]: Is there at least one unit test for SOURCE_DATE_EPOCH set to an empty string?

**Review by vstinner — APPROVED** [review]
LGTM, but I have a few minor comments.

I will not approve this PR since it's the work of @benjaminp and I'm not sure that the change respects his design (PEP 552) (I'm not sure that he agrees with the change).

**Comment by elprans** [comment]
> I will not approve this PR since it's the work of @benjaminp and I'm not sure that the change respects his design (PEP 552) (I'm not sure that he agrees with the change).

Benjamin's [comment on the issue](https://bugs.python.org/issue34022#msg326054):

> I dislike that having SOURCE_DATE_EPOCH set magically changes command line tools behavior. In my view, this problem should be fixed by reverting ccbe5818af2.

Reverting seems like a drastic measure, I think this PR achieves a better balance.

File: Doc/library/py_compile.rst
Comment (elprans) [reply Doc/library/py_compile.rst:60]: PEP 552 actually has no mention of `SOURCE_DATE_EPOCH`, so I'm not sure such a reference would be useful here.

Comment (vstinner) [reply Doc/library/py_compile.rst:60]: invalidation_mode is a new thing coming from the PEP 552. Moreover, I proposed to update the PEP:
https://bugs.python.org/issue34022#msg326638

**Comment by vstinner** [comment]
cc @bmwiedemann who authored PR #5200.

**Comment by vstinner** [comment]
Hum, a few tests still fail with SOURCE_DATE_EPOCH=0:

* test_multiprocessing_main_handling
* test_cmd_line_script
* test_runpy

@elprans: Do you want to fix these tests in the same PR, or do you prefer to fix them in a different PR?

**Comment by elprans** [comment]
@vstinner Separate PR. Those are unrelated to `py_compile` and are due to broken `zipimport`.

**Comment by vstinner** [comment]
@ned-deily, @benjaminp, @elprans: I guess that this change can wait Python 3.7.2, it's not a blocker issue nor a major bug. So I suggest to wait until 3.7.1 final is released before backporting this change to the 3.7 branch.

**Comment by vstinner** [comment]
> @vstinner Separate PR. Those are unrelated to py_compile and are due to broken zipimport.

Perfect. In this case, I see no other reason to no merge your PR :-) I merged it, good job! Thanks @elprans.

**Comment by ned-deily** [comment]
Unless a release manager says otherwise, please don't wait between RC's and finals to merge PRs.  It's better to get them exposed on buildbots.  We'll cherry-picked fixes as necessary into the finals.

**Comment by vstinner** [comment]
> Unless a release manager says otherwise, please don't wait between RC's and finals to merge PRs. It's better to get them exposed on buildbots. We'll cherry-picked fixes as necessary into the finals.

The merged change is a backward incompatible and it mentions the version 3.7.2 in the modified documentation :-)

**Comment by ned-deily** [comment]
The docs can be easily changed for the final.  The question is how much of an impact is it for 3.7.1 users by not having this change in vs having it in..

**Comment by mcepl** [comment]
> @elprans @vstinner Separate PR. Those are unrelated to py_compile and are due to broken zipimport.

Is there such PR somewhere? Or at least a bug?

**Comment by mcepl** [comment]
@elprans thanks

**Pull request merged**
